[WebCore] Optimize Font::applyTransforms

Constellation opened the pull request:
#### 4e349ca18c0f8e55b6bdcd2786755caf68787143
<pre>
[WebCore] Optimize Font::applyTransforms
<a href="https://bugs.webkit.org/show_bug.cgi?id=270406">https://bugs.webkit.org/show_bug.cgi?id=270406</a>
<a href="https://rdar.apple.com/123961009">rdar://123961009</a>

Reviewed by Chris Dumez.

Font::applyTransforms is very slow. While the most of time is used in CoreText, the other part is also using much time!
This patch optimizes it.

1. We add Vector::insertFill function to insert one-item-filling into Vector. GlyphBuffer is
   doing this in a very inefficient way right now: allocating filled Vector and using insertVector.
2. Add size parameter to upconvertedCharacters and use 256 for static Vector size in Font::applyTransforms,
   to avoid unnecessary allocations.
3. LocaleCocoa::canonicalLanguageIdentifierFromString should return RetainPtr&lt;CFStringRef&gt;. We found that
   we are super repeatedly creating CFString when locale is specified because canonicalLanguageIdentifierFromString
   returns AtomString and we convert it to CFString. And this is very slow. Because canonicalLanguageIdentifierFromString
   is only used in this place, we should just return RetainPtr&lt;CFStringRef&gt;. Also we optimized the caching mechanism
   in canonicalLanguageIdentifierFromString to cache the one item out of HashMap since this one-item cache can cover almost
   all cases.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc&gt;::insertFill):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::upconvertedCharacters const):
(WTF::StringView::UpconvertedCharacters&lt;N&gt;::UpconvertedCharacters):
(WTF::StringView::UpconvertedCharacters::UpconvertedCharacters): Deleted.
* Source/WebCore/editing/TextIterator.cpp:
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::makeHole):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::applyTransforms const):
* Source/WebCore/platform/text/cocoa/LocaleCocoa.h:
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::localeCache):
(WebCore::LocaleCocoa::canonicalLanguageIdentifierFromString):
(WebCore::LocaleCocoa::releaseMemory):
(WebCore::canonicalLocaleMap): Deleted.

Canonical link: <a href="https://commits.webkit.org/275676@main">https://commits.webkit.org/275676@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d04a611ea9e6226b3c4576039e2a104e67a70997

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42491 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21509 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44886 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45092 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38608 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/44798 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24741 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18864 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35179 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43065 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18426 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36574 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16121 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16069 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37623 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46567 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/35947 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38669 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37947 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41861 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/42119 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14249 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40469 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18926 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/49128 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18993 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9948 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5737 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18571 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/Vector.h:
```diff
@@ -889,6 +889,7 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     void appendUsingFunctor(size_t, const Functor&);
 
     void insert(size_t position, value_type&& value) { insert<value_type>(position, std::forward<value_type>(value)); }
+    void insertFill(size_t position, const T& value, size_t dataSize);
```
Can this take a std::span<> instead?

(this would be a good candidate for a unit test in Tools/TestWebKitAPI/Tests/WTF/Vector.cpp)

Constellation replied:
Here, we cannot use `std::span` since it requires underlying elements, but this function does not have that. So this is not a span.

vitorroriz commented on Source/WTF/wtf/Vector.h:
```diff
@@ -1633,6 +1634,24 @@ inline void Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::ins
+template<typename T, size_t inlineCapacity, typename OverflowHandler, size_t minCapacity, typename Malloc>
+void Vector<T, inlineCapacity, OverflowHandler, minCapacity, Malloc>::insertFill(size_t position, const T& data, size_t dataSize)
+{
+    ASSERT_WITH_SECURITY_IMPLICATION(position <= size());
```
can position be equal size?

Constellation replied:
Yes. It is append.

vitorroriz commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -112,7 +112,7 @@ class SearchBuffer {
 
     const String m_target;
-    const StringView::UpconvertedCharacters m_targetCharacters;
+    const StringView::UpconvertedCharacters<32> m_targetCharacters;
```
nit: I know you haven't introduced this constant but it might be a good chance to replace it with a named constexpr

Constellation replied:
Make it as just a default `UpconvertedCharacters`.

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -596,8 +596,8 @@ GlyphBufferAdvance Font::applyTransforms(GlyphBuffer& glyphBuffer, unsigned begi
     auto substring = text.substring(beginningStringIndex);
-    auto upconvertedCharacters = substring.upconvertedCharacters();
-    auto localeString = locale.isNull() ? nullptr : LocaleCocoa::canonicalLanguageIdentifierFromString(locale).string().createCFString();
+    auto upconvertedCharacters = substring.upconvertedCharacters<256>();
```
nit: maybe we could have 256 as a named constexpr

Constellation replied:
Done.

cdumez commented on Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
```diff
@@ -245,26 +245,43 @@
 {
-    static NeverDestroyed<CanonicalLocaleMap> canonicalLocaleMap;
-    return canonicalLocaleMap.get();
+    static NeverDestroyed<LocaleCache> localeCache;
```
nit: Would be nice to switch to `MainThreadNeverDestroyed` since you're updating this line.

Constellation replied:
Changed.

cdumez reviewed (approved):
LGTM

Constellation committed:
[WebCore] Optimize Font::applyTransforms
https://bugs.webkit.org/show_bug.cgi?id=270406
rdar://123961009

Reviewed by Chris Dumez.

Font::applyTransforms is very slow. While the most of time is used in CoreText, the other part is also using much time!
This patch optimizes it.

1. We add Vector::insertFill function to insert one-item-filling into Vector. GlyphBuffer is
   doing this in a very inefficient way right now: allocating filled Vector and using insertVector.
2. Add size parameter to upconvertedCharacters and use 256 for static Vector size in Font::applyTransforms,
   to avoid unnecessary allocations.
3. LocaleCocoa::canonicalLanguageIdentifierFromString should return RetainPtr<CFStringRef>. We found that
   we are super repeatedly creating CFString when locale is specified because canonicalLanguageIdentifierFromString
   returns AtomString and we convert it to CFString. And this is very slow. Because canonicalLanguageIdentifierFromString
   is only used in this place, we should just return RetainPtr<CFStringRef>. Also we optimized the caching mechanism
   in canonicalLanguageIdentifierFromString to cache the one item out of HashMap since this one-item cache can cover almost
   all cases.

* Source/WTF/wtf/Vector.h:
(WTF::Malloc>::insertFill):
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::upconvertedCharacters const):
(WTF::StringView::UpconvertedCharacters<N>::UpconvertedCharacters):
(WTF::StringView::UpconvertedCharacters::UpconvertedCharacters): Deleted.
* Source/WebCore/editing/TextIterator.cpp:
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::makeHole):
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::applyTransforms const):
* Source/WebCore/platform/text/cocoa/LocaleCocoa.h:
* Source/WebCore/platform/text/cocoa/LocaleCocoa.mm:
(WebCore::localeCache):
(WebCore::LocaleCocoa::canonicalLanguageIdentifierFromString):
(WebCore::LocaleCocoa::releaseMemory):
(WebCore::canonicalLocaleMap): Deleted.

Canonical link: https://commits.webkit.org/275676@main
